client.go

```diff
@@ -521,6 +527,33 @@ func (c *Client) httpRequest(method string, path string, data []byte, headers ma
 	return response, buf, nil
 }
 
+func (c *Client) host(useCoordinator bool) (*URI, error) {
+	var host *URI
+	if useCoordinator {
+		c.coordinatorLock.RLock()
+		coordinatorURI := c.coordinatorURI
+		c.coordinatorLock.RUnlock()
+		if coordinatorURI == nil {
+			node, err := c.fetchCoordinatorNode()
```

Comment:
I believe that we need to obtain the write lock before calling `fetchCoordinatorNode`, my reasoning is as follows:

multiple routines may call Client.host simultaneously. They might simultaneously obtain a read lock and see that the coordinator is nil, they would then all try to fetch the coordinator node simultaneously which would be wasteful since only one needs to do it. Then they will all try to obtain the write lock to set the coordinator node to what they found, and each one will succeed in succession.

What we should do is obtain the write lock, then *re-check that c.coordinatorURI is still nil*, if it is still nil, then proceed with fetchCoordinator node. If it was set in between releasing the RLock and obtaining the Lock, then we can release the lock and use it.